dask/array/_array_expr/_expr.py

```diff
@@ -163,5 +189,94 @@ def unify_chunks_expr(*args):
                 changed = True
             else:
                 pass
-        arrays.extend([a, i])
+        arrays.append(a)
     return chunkss, arrays, changed
+
+
+class Stack(ArrayExpr):
+    _parameters = ["array", "axis", "meta"]
+
+    @functools.cached_property
+    def args(self):
+        return [self.array] + self.operands[len(self._parameters) :]
+
+    @functools.cached_property
+    def _meta(self):
+        return self.operand("meta")
+
+    @functools.cached_property
+    def chunks(self):
+        n = len(self.args)
+        return (
+            self.array.chunks[: self.axis]
+            + ((1,) * n,)
+            + self.array.chunks[self.axis :]
+        )
+
+    @functools.cached_property
+    def _name(self):
+        return "stack-" + _tokenize_deterministic(*self.operands)
+
+    def _layer(self) -> dict:
+        keys = list(product([self._name], *[range(len(bd)) for bd in self.chunks]))
+        names = [a.name for a in self.args]
+        axis = self.axis
+        ndim = self._meta.ndim - 1
+
+        inputs = [
+            (names[key[axis + 1]],) + key[1 : axis + 1] + key[axis + 2 :]
+            for key in keys
+        ]
+        values = [
+            (
+                getitem,
+                inp,
+                (slice(None, None, None),) * axis
+                + (None,)
+                + (slice(None, None, None),) * (ndim - axis),
+            )
+            for inp in inputs
+        ]
+        return dict(zip(keys, values))
+
+
+class Concatenate(ArrayExpr):
+    _parameters = ["array", "axis", "meta"]
+
+    @functools.cached_property
+    def args(self):
+        return [self.array] + self.operands[len(self._parameters) :]
+
+    @functools.cached_property
+    def _meta(self):
+        return self.operand("meta")
+
+    @functools.cached_property
+    def chunks(self):
+        bds = [a.chunks for a in self.args]
+        chunks = (
+            bds[0][: self.axis]
+            + (sum((bd[self.axis] for bd in bds), ()),)
+            + bds[0][self.axis + 1 :]
+        )
+        return chunks
+
+    @functools.cached_property
+    def _name(self):
+        return "stack-" + _tokenize_deterministic(*self.operands)
+
+    def _layer(self) -> dict:
+        axis = self.axis
+        cum_dims = [0] + list(accumulate(add, [len(a.chunks[axis]) for a in self.args]))
+        keys = list(product([self._name], *[range(len(bd)) for bd in self.chunks]))
+        names = [a.name for a in self.args]
+
+        values = [
+            (names[bisect(cum_dims, key[axis + 1]) - 1],)
+            + key[1 : axis + 1]
+            + (key[axis + 1] - cum_dims[bisect(cum_dims, key[axis + 1]) - 1],)
+            + key[axis + 2 :]
+            for key in keys
+        ]
```

Comment:
I assume this is copied?? Are we actually running that many bisects at graph materialization??

Reply:
yes this is a complete copy